.github/workflows/test-integration.yml

```diff
@@ -0,0 +1,77 @@
+name: Integration
+on:
+  pull_request:
+  push:
+    paths:
+      - '**.go'
+      - '**/go.mod'
+      - '**/go.sum'
+      - '**/testdata/**'
+    branches:
+      - main
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  pre-test:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v3.4.0
+      - name: Finding files and store to output
+        id: set-matrix
+        run: echo "matrix=$({ cd integration && find . -type d ! -name testdata -maxdepth 1 -print; } | tail -n +2 | cut -c 3- | jq -R . | jq -cs .)" >> $GITHUB_OUTPUT
+
+  integration:
+    name: test ${{ matrix.test-path }}
+    runs-on: ubuntu-latest
+    needs: pre-test
+    if: fromJSON(needs.pre-test.outputs.matrix)[0] != null
+    continue-on-error: false
+    strategy:
+      fail-fast: true
+      matrix:
+        test-path: ${{fromJson(needs.pre-test.outputs.matrix)}}
+    steps:
+      - uses: actions/checkout@v3.4.0
+      - uses: technote-space/get-diff-action@v6.1.2
+        with:
+          PATTERNS: |
+            ${{ matrix.test-path }}/**
+            !${{ matrix.test-path }}/README.md
+            integration/go.mod
+            integration/go.sum
+            integration/${{ matrix.test-path }}/**
+
+      - uses: actions/cache@v3
+        if: env.GIT_DIFF
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
```

Comment:
This one can be removed because `actions/setup-go@v4` already handles Go caching:

```suggestion
```